builtins/web/fetch/request-response.cpp

```diff
@@ -2291,6 +2291,91 @@ bool Response::redirect(JSContext *cx, unsigned argc, Value *vp) {
 //   return true;
 // }
 
+/// https://fetch.spec.whatwg.org/#dom-response-clone
+bool Response::clone(JSContext *cx, unsigned argc, JS::Value *vp) {
+  // If this is unusable, then throw a TypeError.
+  METHOD_HEADER(0);
+
+  // To clone a response response, run these steps:
+  // 1. If response is a filtered response, then return a new identical filtered response whose internal response is a clone of response’s internal response.
+  RootedObject new_response(cx, create(cx));
+  if (!new_response) {
+    return false;
+  }
+
+  init_slots(new_response);
+
+  // 2. Let newResponse be a copy of response, except for its body.
+  RootedValue cloned_headers_val(cx, JS::NullValue());
+  RootedObject headers(cx, RequestOrResponse::maybe_headers(self));
+  if (headers) {
+    RootedValue headers_val(cx, ObjectValue(*headers));
+    JSObject *cloned_headers = Headers::create(cx, headers_val, Headers::guard(headers));
+    if (!cloned_headers) {
+      return false;
+    }
+    cloned_headers_val.set(ObjectValue(*cloned_headers));
+  } else if (RequestOrResponse::maybe_handle(self)) {
+    auto handle = RequestOrResponse::headers_handle_clone(cx, self);
+    JSObject *cloned_headers =
+        Headers::create(cx, handle.release(),
+                        RequestOrResponse::is_incoming(self) ? Headers::HeadersGuard::Immutable
+                                                             : Headers::HeadersGuard::Response);
+    if (!cloned_headers) {
+      return false;
+    }
+    cloned_headers_val.set(ObjectValue(*cloned_headers));
+  }
```

Comment:
I think you should be able to use `RequestOrResponse::headers_handle_clone` for this, reducing complexity a bunch. If that's not quite correct to use here, perhaps it can instead be tweaked?

Reply:
I was unable to work out a simple way to use `RequestOrResponse::headers_handle_clone` for this, however the solution I've added in a recent commit that constructs a `RootedObject` based on `RequestOrResponse::headers` appears to do the trick. I'd love an extra set of eyes on that approach though as I'm still wrapping my head around internal object apis.
```c++
RootedObject cloned_headers(cx, RequestOrResponse::headers(cx, self));
if (!cloned_headers) {
  return false;
}
```

Reply:
I gave it another look over and was able to get `headers_handle_clone` to work for this. I'll chalk it up to my own ignorance regarding the codebase.  I've mirrored the changes to `Request::clone` as well and wrote up some changes to the tests to check that the headers are properly cloned rather than shared.

Once any further changes are worked through, do you prefer keeping the individual commits in this PR or would a force-push of squashed commits be preferred?